README.md
# binance-aio 0.0.3

**Announcement:**`binance-aio` has been replaced by a new library [`cryptoxlib-aio`](https://github.com/nardew/cryptoxlib-aio). `cryptoxlib-aio` offers the very same functionality as `binance-aio` but on top it provides access to multiple cryptoexchanges and other (mostly technical) new features. You can keep using `binance-aio` but please note no new features/bugfixes will be implemented. We recommend to migrate to `cryptoxlib-aio`.

----

[![](https://img.shields.io/badge/python-3.6-blue.svg)](https://www.python.org/downloads/release/python-365/) [![](https://img.shields.io/badge/python-3.7-blue.svg)](https://www.python.org/downloads/release/python-374/)

`binance-aio` is a Python library providing access to [binance crypto exchange](https://www.binance.com/en). Library implements binance's REST API as well as websockets.

`binance-aio` is designed as an asynchronous library utilizing modern features of Python and of supporting asynchronous libraries (mainly [async websockets](https://websockets.readthedocs.io/en/stable/) and [aiohttp](https://aiohttp.readthedocs.io/en/stable/)).

For changes see [CHANGELOG](https://github.com/nardew/binance-aio/blob/master/CHANGELOG.md).

### Features
- access to limited binance's REST API (account details, market data, order management, ...) and websockets (account feed, market data feed, orderbook feed, ...). Missing REST calls and websocket streams will be added on request and based on our availability.
- channels bundled in one or multiple websockets processed in parallel 
- lean architecture setting ground for the future extensions and customizations
- fully asynchronous design aiming for the best performance

### Installation
```bash
pip install binance-aio
```

### Prerequisites

Due to dependencies and Python features used by the library please make sure you use Python `3.6` or `3.7`.

Before starting using `binance-aio`, it is necessary to take care of:
1. downloading your Binance API and SECRET key from your binance account
1. generating your certificate that will be used to secure SSL connections. Certificate `certificate.pem` can be generated easily by
```bash
openssl req -newkey rsa:2048 -new -nodes -x509 -days 3650 -keyout key.pem -out certificate.pem
```

### Examples
#### REST API
```python
import asyncio
import pathlib
import logging
import os
from datetime import datetime

from binance.BinanceClient import BinanceClient
from binance.Pair import Pair
from binance.enums import OrderSide, TimeInForce, OrderResponseType

LOG = logging.getLogger("binance")
LOG.setLevel(logging.DEBUG)
LOG.addHandler(logging.StreamHandler())

print(f"Available loggers: {[name for name in logging.root.manager.loggerDict]}\n")

async def account_update(response : dict) -> None:
	print(f"Callback {account_update.__name__}: [{response}]")

async def order_book_update(response : dict) -> None:
	print(f"Callback {order_book_update.__name__}: [{response}]")

async def trade_update(response : dict) -> None:
	local_timestamp_ms = int(datetime.now().timestamp() * 1000)
	server_timestamp_ms = response['E']
	print(f"Trade update timestamp diff [ms]: {local_timestamp_ms - server_timestamp_ms}")

async def orderbook_ticker_update(response : dict) -> None:
	print(f"Callback {orderbook_ticker_update.__name__}: [{response}]")

async def run():
	print("STARTING BINANCE CLIENT\n")

	# to generate a certificate use 'openssl req -newkey rsa:2048 -new -nodes -x509 -days 3650 -keyout key.pem -out certificate.pem'
	certificate_path = pathlib.Path(__file__).with_name("certificate.pem")

	# to retrieve your API/SEC key go to your binance website, create the keys and store them in APIKEY/SECKEY
	# environment variables
	api_key = os.environ['APIKEY']
	sec_key = os.environ['SECKEY']

	client = BinanceClient(certificate_path, api_key, sec_key)

	# REST api calls
	print("REST API")

	print("\nPing:")
	await client.ping()

	print("\nServer time:")
	await client.get_time()

	print("\nExchange info:")
	await client.get_exchange_info()

	print("\nBest order book ticker:")
	await client.get_best_orderbook_ticker(pair = Pair('ETH', 'BTC'))

	print("\nAccount:")
	await client.get_account(recv_window_ms = 5000)

	print("\nCreate limit order:")
	await client.create_limit_order(Pair("ETH", "BTC"), OrderSide.BUY, "1", "0", time_in_force = TimeInForce.GOOD_TILL_CANCELLED,
	                                new_order_response_type = OrderResponseType.FULL)

	print("\nDelete order:")
	await client.delete_order(pair = Pair('ETH', 'BTC'), order_id = "1")

	await client.close()

if __name__ == "__main__":
	asyncio.run(run())
```

#### WEBSOCKETS
```python
import asyncio
import pathlib
import logging
import os
from datetime import datetime

from binance.BinanceClient import BinanceClient
from binance.Pair import Pair
from binance.subscriptions import BestOrderBookTickerSubscription, TradeSubscription, AccountSubscription

LOG = logging.getLogger("binance")
LOG.setLevel(logging.DEBUG)
LOG.addHandler(logging.StreamHandler())

print(f"Available loggers: {[name for name in logging.root.manager.loggerDict]}\n")

async def account_update(response : dict) -> None:
	print(f"Callback {account_update.__name__}: [{response}]")

async def order_book_update(response : dict) -> None:
	print(f"Callback {order_book_update.__name__}: [{response}]")

async def trade_update(response : dict) -> None:
	local_timestamp_ms = int(datetime.now().timestamp() * 1000)
	server_timestamp_ms = response['E']
	print(f"Trade update timestamp diff [ms]: {local_timestamp_ms - server_timestamp_ms}")

async def orderbook_ticker_update(response : dict) -> None:
	print(f"Callback {orderbook_ticker_update.__name__}: [{response}]")

async def run():
	print("STARTING BINANCE CLIENT\n")

	# to generate a certificate use 'openssl req -newkey rsa:2048 -new -nodes -x509 -days 3650 -keyout key.pem -out certificate.pem'
	certificate_path = pathlib.Path(__file__).with_name("certificate.pem")

	# to retrieve your API/SEC key go to your binance website, create the keys and store them in API_KEY/SEC_KEY
	# environment variables
	api_key = os.environ['APIKEY']
	sec_key = os.environ['SECKEY']

	client = BinanceClient(certificate_path, api_key, sec_key)

	# Websockets
	print("\nWEBSOCKETS\n")

	print("\nCreate listen key:")
	listen_key = await client.get_listen_key()

	# Bundle several subscriptions into a single websocket
	client.compose_subscriptions([
		BestOrderBookTickerSubscription(callbacks = [orderbook_ticker_update]),
		TradeSubscription(pair = Pair('ETH', 'BTC'), callbacks = [trade_update])
	])

	# Bundle another subscriptions into a separate websocket
	print(listen_key)
	client.compose_subscriptions([
		AccountSubscription(client, callbacks = [account_update])
	])

	# Execute all websockets asynchronously
	await client.start_subscriptions()

	await client.close()

if __name__ == "__main__":
	asyncio.run(run())
```

All examples can be found in `client-example/client.py` in the GitHub repository.

### Support

If you like the library and you feel like you want to support its further development, enhancements and bugfixing, then it will be of great help and most appreciated if you:
- file bugs, proposals, pull requests, ...
- spread the word
- donate an arbitrary tip
  * BTC: 15JUgVq3YFoPedEj5wgQgvkZRx5HQoKJC4
  * ETH: 0xf29304b6af5831030ba99aceb290a3a2129b993d
  * ADA: DdzFFzCqrhshyLV3wktXFvConETEr9mCfrMo9V3dYz4pz6yNq9PjJusfnn4kzWKQR91pWecEbKoHodPaJzgBHdV2AKeDSfR4sckrEk79
  * XRP: rhVWrjB9EGDeK4zuJ1x2KXSjjSpsDQSaU6 **+ tag** 599790141

### Contact

If you feel you want to get in touch, then please

- preferably use Github Issues, or
- send an e-mail to <img src="http://safemail.justlikeed.net/e/160b5d5ea6878dd3d893b9f9fecb6ed7.png" border="0" align="absbottom">

### Affiliation

In case you are interested in an automated trading bot, check out our other project [creten](https://github.com/nardew/creten).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Pending release]

## [0.0.3] - 2020-03-31

### Changed

- New documentation

## 0.0.1 - 2020-01-30

### Added

- Initial version supporting limited set of REST calls and websockets

[Pending release]: https://github.com/nardew/binance-aio/compare/0.0.2...HEAD
[0.0.2]: https://github.com/nardew/bitpanda-aio/compare/0.0.1...0.0.2

binance/enums.py
import enum

class RestCallType(enum.Enum):
	GET = enum.auto()
	POST = enum.auto()
	DELETE = enum.auto()
	PUT = enum.auto()

class OrderSide(enum.Enum):
	BUY = "BUY"
	SELL = "SELL"

class OrderType(enum.Enum):
	LIMIT = 'LIMIT'
	MARKET = 'MARKET'
	STOP_LOSS = 'STOP_LOSS'
	STOP_LOSS_LIMIT = 'STOP_LOSS_LIMIT'
	TAKE_PROFIT = 'TAKE_PROFIT'
	TAKE_PROFIT_LIMIT = 'TAKE_PROFIT_LIMIT'
	LIMIT_MAKER = 'LIMIT_MAKER'

class DepthLimit(enum.Enum):
	L_5 = '5'
	L_10 = '10'
	L_20 = '20'
	L_50 = '50'
	L_100 = '100'
	L_500 = '500'
	L_1000 = '1000'
	L_5000 = '5000'

class CandelstickInterval(enum.Enum):
	I_1MIN = '1m'
	I_3MIN = '3m'
	I_5MIN = '5m'
	I_15MIN = '15m'
	I_30MIN = '30m'
	I_1H = '1h'
	I_2H = '2h'
	I_4H = '4h'
	I_6H = '6h'
	I_8H = '8h'
	I_12H = '12h'
	I_1D = '1d'
	I_3D = '3d'
	I_1W = '1w'
	I_1MONTH = '1M'

class OrderResponseType(enum.Enum):
	ACT = "ACK"
	RESULT = "RESULT"
	FULL = "FULL"

class TimeInForce(enum.Enum):
	GOOD_TILL_CANCELLED = "GTC"
	IMMEDIATE_OR_CANCELLED = "IOC"
	FILL_OR_KILL = "FOK"



binance/BinanceClient.py
import asyncio
import aiohttp
import hmac
import hashlib
import ssl
import logging
import datetime
import json
from typing import List, Optional

from binance.Pair import Pair
from binance.subscriptions import Subscription, SubscriptionMgr
from binance import enums
from binance.Timer import Timer
from binance.BinanceException import BinanceException

LOG = logging.getLogger(__name__)

class BinanceClient(object):
	REST_API_URI = "https://api.binance.com/api/v3/"

	def __init__(self, certificate_path : str = None, api_key : str = None, sec_key : str = None,
	             api_trace_log : bool = False) -> None:
		self.api_key = api_key
		self.sec_key = sec_key
		self.api_trace_log = api_trace_log

		self.rest_session = None

		self.ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
		self.ssl_context.load_verify_locations(certificate_path)

		self.subscription_sets = []

	async def ping(self) -> dict:
		return await self._create_get("ping")

	async def get_exchange_info(self) -> dict:
		return await self._create_get("exchangeInfo")

	async def get_time(self) -> dict:
		return await self._create_get("time")

	async def get_orderbook(self, pair : Pair, limit : enums.DepthLimit = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": pair,
		})

		if limit:
			params['limit'] = limit.value

		return await self._create_get("depth", params = params)

	async def get_trades(self, pair : Pair, limit : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": pair,
			"limit": limit
		})

		return await self._create_get("trades", params = params)

	async def get_historical_trades(self, pair : Pair, limit : int = None, from_id : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": pair,
			"limit": limit,
			"fromId": from_id
		})

		return await self._create_get("historicalTrades", params = params, headers = self._get_header_api_key())

	async def get_aggregate_trades(self, pair : Pair, limit : int = None, from_id : int = None, start_tmstmp_ms : int = None, end_tmstmp_ms : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": pair,
			"limit": limit,
			"fromId": from_id,
			"startTime": start_tmstmp_ms,
			"endTime": end_tmstmp_ms
		})

		return await self._create_get("aggTrades", params = params)

	async def get_candelsticks(self, pair : Pair, limit : int = None, interval : enums.CandelstickInterval = None, start_tmstmp_ms : int = None, end_tmstmp_ms : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": pair,
			"limit": limit,
			"startTime": start_tmstmp_ms,
			"endTime": end_tmstmp_ms
		})

		if interval:
			params['interval'] = interval.value

		return await self._create_get("klines", params = params)

	async def get_average_price(self, pair : Pair) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": pair
		})

		return await self._create_get("avgPrice", params = params)

	async def get_24h_price_ticker(self, pair : Pair = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": pair,
		})

		return await self._create_get("ticker/24hr", params = params)

	async def get_price_ticker(self, pair : Pair = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": pair,
		})

		return await self._create_get("ticker/price", params = params)

	async def get_best_orderbook_ticker(self, pair : Optional[Pair] = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": pair,
		})

		return await self._create_get("ticker/bookTicker", headers = self._get_header_api_key(), params = params)

	async def create_order(self, pair : Pair, side : enums.OrderSide, type : enums.OrderType,
	                             quantity : str,
	                             price : str = None,
								 stop_price: str = None,
	                             quote_order_quantity: str = None,
	                             time_in_force: enums.TimeInForce = None,
	                             new_client_order_id : str = None,
	                             iceberg_quantity : str = None,
	                             new_order_response_type : enums.OrderResponseType = None,
	                             recv_window_ms : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": str(pair),
			"side": side.value,
			"type": type.value,
			"quantity": quantity,
			"quoteOrderQty": quote_order_quantity,
			"price": price,
			"stopPrice": stop_price,
			"newClientOrderId": new_client_order_id,
			"icebergQty": iceberg_quantity,
			"recvWindow": recv_window_ms,
			"timestamp": self._get_current_timestamp_ms()
		})

		if time_in_force:
			params['timeInForce'] = time_in_force.value

		if new_order_response_type:
			params['newOrderRespType'] = new_order_response_type.value

		return await self._create_post("order", params = params, headers = self._get_header_api_key(), signed = True)

	async def create_test_order(self, pair : Pair, side : enums.OrderSide, type : enums.OrderType,
	                             quantity : str,
	                             price : str = None,
								 stop_price: str = None,
	                             quote_order_quantity: str = None,
	                             time_in_force: enums.TimeInForce = None,
	                             new_client_order_id : str = None,
	                             iceberg_quantity : str = None,
	                             new_order_response_type : enums.OrderResponseType = None,
	                             recv_window_ms : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": str(pair),
			"side": side.value,
			"type": type.value,
			"quantity": quantity,
			"quoteOrderQty": quote_order_quantity,
			"price": price,
			"stopPrice": stop_price,
			"newClientOrderId": new_client_order_id,
			"icebergQty": iceberg_quantity,
			"recvWindow": recv_window_ms,
			"timestamp": self._get_current_timestamp_ms()
		})

		if time_in_force:
			params['timeInForce'] = time_in_force.value

		if new_order_response_type:
			params['newOrderRespType'] = new_order_response_type.value

		return await self._create_post("order/test", params = params, headers = self._get_header_api_key(), signed = True)

	async def get_order(self, pair : Pair, order_id : int = None, orig_client_order_id : int = None, recv_window_ms : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": str(pair),
			"orderId": order_id,
			"origClientOrderId": orig_client_order_id,
			"recvWindow": recv_window_ms,
			"timestamp": self._get_current_timestamp_ms()
		})

		return await self._create_get("order", params = params, headers = self._get_header_api_key(), signed = True)

	async def cancel_order(self, pair : Pair, order_id : str = None, orig_client_order_id : str = None,
	                       new_client_order_id : str = None, recv_window_ms : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": str(pair),
			"orderId": order_id,
			"origClientOrderId": orig_client_order_id,
			"newClientOrderId": new_client_order_id,
			"recvWindow": recv_window_ms,
			"timestamp": self._get_current_timestamp_ms()
		})

		return await self._create_delete("order", params = params, headers = self._get_header_api_key(), signed = True)

	async def get_open_orders(self, pair : Pair = None, recv_window_ms : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": str(pair),
			"recvWindow": recv_window_ms,
			"timestamp": self._get_current_timestamp_ms()
		})

		return await self._create_get("openOrders", params = params, headers = self._get_header_api_key(), signed = True)

	async def get_all_orders(self, pair : Pair, order_id : int = None, limit : int = None, start_tmstmp_ms : int = None, end_tmstmp_ms : int = None, recv_window_ms : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": str(pair),
			"orderId": order_id,
			"startTime": start_tmstmp_ms,
			"endTime": end_tmstmp_ms,
			"limit": limit,
			"recvWindow": recv_window_ms,
			"timestamp": self._get_current_timestamp_ms()
		})

		return await self._create_get("allOrders", params = params, headers = self._get_header_api_key(), signed = True)

	async def create_oco_order(self, pair : Pair, side : enums.OrderSide,
	                           quantity : str,
	                           price: str,
	                           stop_price: str,
	                           limit_client_order_id: str = None,
	                           list_client_order_id: str = None,
	                           limit_iceberg_quantity: str = None,
	                           stop_client_order_id: str = None,
	                           stop_limit_price: str = None,
	                           stop_iceberg_quantity: str = None,
	                           stop_limit_time_in_force: enums.TimeInForce = None,
	                           new_order_response_type : enums.OrderResponseType = None,
	                           recv_window_ms : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": str(pair),
			"side": side.value,
			"quantity": quantity,
			"listClientOrderId": list_client_order_id,
			"limitClientOrderId": limit_client_order_id,
			"price": price,
			"stopClientOrderId": stop_client_order_id,
			"stopPrice": stop_price,
			"stopLimitPrice": stop_limit_price,
			"stopIcebergQty": stop_iceberg_quantity,
			"limitIcebergQty": limit_iceberg_quantity,
			"recvWindow": recv_window_ms,
			"timestamp": self._get_current_timestamp_ms()
		})

		if stop_limit_time_in_force:
			params['stopLimitTimeInForce'] = stop_limit_time_in_force.value

		if new_order_response_type:
			params['newOrderRespType'] = new_order_response_type.value

		return await self._create_post("order/oco", params = params, headers = self._get_header_api_key(), signed = True)

	async def cancel_oco_order(self, pair : Pair, order_list_id : str = None, list_client_order_id : str = None,
	                       new_client_order_id : str = None, recv_window_ms : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": str(pair),
			"orderListId": order_list_id,
			"listClientOrderId": list_client_order_id,
			"newClientOrderId": new_client_order_id,
			"recvWindow": recv_window_ms,
			"timestamp": self._get_current_timestamp_ms()
		})

		return await self._create_delete("orderList", params = params, headers = self._get_header_api_key(), signed = True)

	async def get_oco_order(self, order_list_id : int = None, orig_client_order_id : int = None, recv_window_ms : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"orderListId": order_list_id,
			"origClientOrderId": orig_client_order_id,
			"recvWindow": recv_window_ms,
			"timestamp": self._get_current_timestamp_ms()
		})

		return await self._create_get("orderList", params = params, headers = self._get_header_api_key(), signed = True)

	async def get_all_oco_orders(self, from_id : int = None, limit : int = None, start_tmstmp_ms : int = None, end_tmstmp_ms : int = None, recv_window_ms : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"fromId": from_id,
			"startTime": start_tmstmp_ms,
			"endTime": end_tmstmp_ms,
			"limit": limit,
			"recvWindow": recv_window_ms,
			"timestamp": self._get_current_timestamp_ms()
		})

		return await self._create_get("allOrderList", params = params, headers = self._get_header_api_key(), signed = True)

	async def get_open_oco_orders(self, recv_window_ms : int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"recvWindow": recv_window_ms,
			"timestamp": self._get_current_timestamp_ms()
		})

		return await self._create_get("openOrderList", params = params, headers = self._get_header_api_key(), signed = True)

	async def get_account(self, recv_window_ms: Optional[int] = None) -> dict:
		params = BinanceClient._clean_request_params({
			"recvWindow": recv_window_ms,
			"timestamp": self._get_current_timestamp_ms()
		})

		return await self._create_get("account", headers = self._get_header_api_key(), params = params, signed = True)

	async def get_account_trades(self, pair: Pair, limit: int = None, from_id: int = None,
	                               start_tmstmp_ms: int = None, end_tmstmp_ms: int = None) -> dict:
		params = BinanceClient._clean_request_params({
			"symbol": pair,
			"limit": limit,
			"fromId": from_id,
			"startTime": start_tmstmp_ms,
			"endTime": end_tmstmp_ms,
			"timestamp": self._get_current_timestamp_ms()
		})

		return await self._create_get("myTrades", params = params, headers = self._get_header_api_key(), signed = True)

	async def get_listen_key(self):
		return await self._create_post("userDataStream", headers = self._get_header_api_key())

	def compose_subscriptions(self, subscriptions : List[Subscription]) -> None:
		self.subscription_sets.append(subscriptions)

	async def start_subscriptions(self) -> None:
		if len(self.subscription_sets):
			done, pending = await asyncio.wait(
				[asyncio.create_task(SubscriptionMgr(subscriptions, self.api_key, self.ssl_context).run()) for subscriptions in self.subscription_sets],
				return_when = asyncio.FIRST_EXCEPTION
			)
			for task in done:
				try:
					task.result()
				except Exception as e:
					LOG.exception(f"Unrecoverable exception occurred while processing messages: {e}")
					LOG.info("All websockets scheduled for shutdown")
					for task in pending:
						if not task.cancelled():
							task.cancel()
		else:
			raise Exception("ERROR: There are no subscriptions to be started.")

	async def close(self) -> None:
		session = self._get_rest_session()
		if session is not None:
			await session.close()

	async def _create_get(self, resource : str, params : dict = None, headers : dict = None, signed : bool = False) -> dict:
		return await self._create_rest_call(enums.RestCallType.GET, resource, None, params, headers, signed)

	async def _create_post(self, resource : str, data : dict = None, params : dict = None, headers : dict = None, signed : bool = False) -> dict:
		return await self._create_rest_call(enums.RestCallType.POST, resource, data, params, headers, signed)

	async def _create_delete(self, resource : str, params : dict = None, headers : dict = None, signed : bool = False) -> dict:
		return await self._create_rest_call(enums.RestCallType.DELETE, resource, None, params, headers, signed)

	async def _create_put(self, resource : str, params : dict = None, headers : dict = None, signed : bool = False) -> dict:
		return await self._create_rest_call(enums.RestCallType.PUT, resource, None, params, headers, signed)

	async def _create_rest_call(self, rest_call_type : enums.RestCallType, resource : str, data : dict = None, params : dict = None, headers : dict = None, signed : bool = False) -> dict:
		with Timer('RestCall'):
			# add signature into parameters
			if signed:
				params = {} if params is None else params
				params['signature'] = self._get_signature(params, data)

			if rest_call_type == enums.RestCallType.GET:
				rest_call = self._get_rest_session().get(BinanceClient.REST_API_URI + resource, json = data, params = params, headers = headers, ssl = self.ssl_context)
			elif rest_call_type == enums.RestCallType.POST:
				rest_call = self._get_rest_session().post(BinanceClient.REST_API_URI + resource, json = data, params = params, headers = headers, ssl = self.ssl_context)
			elif rest_call_type == enums.RestCallType.DELETE:
				rest_call = self._get_rest_session().delete(BinanceClient.REST_API_URI + resource, json = data, params = params, headers = headers, ssl = self.ssl_context)
			elif rest_call_type == enums.RestCallType.PUT:
				rest_call = self._get_rest_session().put(BinanceClient.REST_API_URI + resource, json = data, params = params, headers = headers, ssl = self.ssl_context)
			else:
				raise Exception(f"Unsupported REST call type {rest_call_type}.")

			LOG.debug(f"> rest type [{rest_call_type.name}], resource [{resource}], params [{params}], headers [{headers}], data [{data}]")
			async with rest_call as response:
				status_code = response.status
				response_body = await response.text()

				LOG.debug(f"<: status [{status_code}], response [{response_body}]")

				if str(status_code)[0] != '2':
					raise BinanceException(f"<: status [{status_code}], response [{response_body}]")

				if len(response_body) > 0:
					response_body = json.loads(response_body)

				return {
					"status_code": status_code,
					"response": response_body
				}

	def _get_rest_session(self) -> aiohttp.ClientSession:
		if self.rest_session is not None:
			return self.rest_session

		if self.api_trace_log:
			trace_config = aiohttp.TraceConfig()
			trace_config.on_request_start.append(BinanceClient._on_request_start)
			trace_config.on_request_end.append(BinanceClient._on_request_end)
			trace_configs = [trace_config]
		else:
			trace_configs = None

		self.rest_session = aiohttp.ClientSession(trace_configs=trace_configs)

		return self.rest_session

	def _get_header_api_key(self):
		header = {
			'Accept': 'application/json',
			"X-MBX-APIKEY": self.api_key
		}

		return header

	@staticmethod
	def _clean_request_params(params : dict) -> dict:
		res = {}
		for key, value in params.items():
			if value is not None:
				res[key] = str(value)

		return res

	async def _on_request_start(session, trace_config_ctx, params) -> None:
		LOG.debug(f"> Context: {trace_config_ctx}")
		LOG.debug(f"> Params: {params}")

	async def _on_request_end(session, trace_config_ctx, params) -> None:
		LOG.debug(f"< Context: {trace_config_ctx}")
		LOG.debug(f"< Params: {params}")

	@staticmethod
	def _get_current_timestamp_ms() -> int:
		return int(datetime.datetime.now(tz = datetime.timezone.utc).timestamp() * 1000)

	def _get_signature(self, params : dict, data : dict) -> str:
		params_string = ""
		data_string = ""

		if params is not None:
			params_string = '&'.join([f"{key}={val}" for key, val in params.items()])

		if data is not None:
			data_string = '&'.join(["{}={}".format(param[0], param[1]) for param in data])

		m = hmac.new(self.sec_key.encode('utf-8'), (params_string+data_string).encode('utf-8'), hashlib.sha256)
		return m.hexdigest()

client-example/client.py
import asyncio
import pathlib
import logging
import os
from datetime import datetime

from binance.BinanceClient import BinanceClient
from binance.Pair import Pair
from binance.subscriptions import BestOrderBookTickerSubscription, TradeSubscription, AccountSubscription
from binance import enums
from binance.BinanceException import BinanceException

LOG = logging.getLogger("binance")
LOG.setLevel(logging.DEBUG)
LOG.addHandler(logging.StreamHandler())

print(f"Available loggers: {[name for name in logging.root.manager.loggerDict]}\n")

async def account_update(response : dict) -> None:
	print(f"Callback {account_update.__name__}: [{response}]")

async def order_book_update(response : dict) -> None:
	print(f"Callback {order_book_update.__name__}: [{response}]")

async def trade_update(response : dict) -> None:
	local_timestamp_ms = int(datetime.now().timestamp() * 1000)
	server_timestamp_ms = response['E']
	print(f"Trade update timestamp diff [ms]: {local_timestamp_ms - server_timestamp_ms}")

async def orderbook_ticker_update(response : dict) -> None:
	print(f"Callback {orderbook_ticker_update.__name__}: [{response}]")

async def run():
	print("STARTING BINANCE CLIENT\n")

	# to generate a certificate use 'openssl req -newkey rsa:2048 -new -nodes -x509 -days 3650 -keyout key.pem -out certificate.pem'
	certificate_path = pathlib.Path(__file__).with_name("certificate.pem")

	# to retrieve your API/SEC key go to your binance website, create the keys and store them in APIKEY/SECKEY
	# environment variables
	api_key = os.environ['APIKEY']
	sec_key = os.environ['SECKEY']

	client = BinanceClient(certificate_path, api_key, sec_key)

	# REST api calls
	print("REST API")

	print("\nPing:")
	await client.ping()

	print("\nServer time:")
	await client.get_time()

	print("\nExchange info:")
	await client.get_exchange_info()

	print("\nOrder book:")
	await client.get_orderbook(pair = Pair('ETH', 'BTC'), limit = enums.DepthLimit.L_5)

	print("\nTrades:")
	await client.get_trades(pair=Pair('ETH', 'BTC'), limit = 5)

	print("\nHistorical trades:")
	await client.get_historical_trades(pair=Pair('ETH', 'BTC'), limit = 5)

	print("\nAggregate trades:")
	await client.get_aggregate_trades(pair=Pair('ETH', 'BTC'), limit = 5)

	print("\nCandelsticks:")
	await client.get_candelsticks(pair=Pair('ETH', 'BTC'), interval = enums.CandelstickInterval.I_1D, limit=5)

	print("\nAverage price:")
	await client.get_average_price(pair = Pair('ETH', 'BTC'))

	print("\n24hour price ticker:")
	await client.get_24h_price_ticker(pair = Pair('ETH', 'BTC'))

	print("\nProce ticker:")
	await client.get_price_ticker(pair = Pair('ETH', 'BTC'))

	print("\nBest order book ticker:")
	await client.get_best_orderbook_ticker(pair = Pair('ETH', 'BTC'))

	print("\nCreate test market order:")
	await client.create_test_order(Pair("ETH", "BTC"), side = enums.OrderSide.BUY, type = enums.OrderType.MARKET,
	                          quantity = "1",
	                          new_order_response_type = enums.OrderResponseType.FULL)

	print("\nCreate limit order:")
	try:
		await client.create_order(Pair("ETH", "BTC"), side = enums.OrderSide.BUY, type = enums.OrderType.LIMIT,
		                          quantity = "1",
		                          price = "0",
		                          time_in_force = enums.TimeInForce.GOOD_TILL_CANCELLED,
		                          new_order_response_type = enums.OrderResponseType.FULL)
	except BinanceException as e:
		print(e)

	print("\nCancel order:")
	try:
		await client.cancel_order(pair = Pair('ETH', 'BTC'), order_id = "1")
	except BinanceException as e:
		print(e)

	print("\nGet order:")
	try:
		await client.get_order(pair = Pair('ETH', 'BTC'), order_id = 1)
	except BinanceException as e:
		print(e)

	print("\nGet open orders:")
	await client.get_open_orders(pair = Pair('ETH', 'BTC'))

	print("\nGet all orders:")
	await client.get_all_orders(pair = Pair('ETH', 'BTC'))

	print("\nCreate OCO order:")
	try:
		await client.create_oco_order(Pair("ETH", "BTC"), side = enums.OrderSide.BUY,
		                          quantity = "1",
		                          price = "0",
		                          stop_price = "0",
		                          new_order_response_type = enums.OrderResponseType.FULL)
	except BinanceException as e:
		print(e)

	print("\nCancel OCO order:")
	try:
		await client.cancel_oco_order(pair = Pair('ETH', 'BTC'), order_list_id = "1")
	except BinanceException as e:
		print(e)

	print("\nGet OCO order:")
	try:
		await client.get_oco_order(order_list_id = 1)
	except BinanceException as e:
		print(e)

	print("\nGet open OCO orders:")
	await client.get_open_oco_orders()

	print("\nGet all OCO orders:")
	await client.get_all_oco_orders()

	print("\nAccount:")
	await client.get_account(recv_window_ms = 5000)

	print("\nAccount trades:")
	await client.get_account_trades(pair = Pair('ETH', 'BTC'))

	# Websockets
	print("\nWEBSOCKETS\n")

	print("\nCreate listen key:")
	listen_key = await client.get_listen_key()

	# Bundle several subscriptions into a single websocket
	client.compose_subscriptions([
		BestOrderBookTickerSubscription(callbacks = [orderbook_ticker_update]),
		TradeSubscription(pair = Pair('ETH', 'BTC'), callbacks = [trade_update])
	])

	# Bundle another subscriptions into a separate websocket
	print(listen_key)
	client.compose_subscriptions([
		AccountSubscription(client, callbacks = [account_update])
	])

	# Execute all websockets asynchronously
	#await client.start_subscriptions()

	await client.close()

if __name__ == "__main__":
	asyncio.run(run())
